pkg/node/provisioner.go

```diff
@@ -1,40 +1,131 @@
+	if err := os.MkdirAll(p.volPath(volID), 0750); err != nil {
+		return status.Errorf(codes.Internal, "Publish Volume Failed: %v", err)
+	}
+	return nil
 }
 
-func Initialize(path string) {
-	provisioner.Path = path
+func (p Provisioner) removeDir(volID string) error {
+	if err := os.RemoveAll(p.volPath(volID)); err != nil {
+		return status.Errorf(codes.Internal, "Publish Volume Failed: %v", err)
+	}
+	return nil
 }
 
-func Provision(volumeID string) (string, error) {
-	provisionerLock.Lock()
-	defer provisionerLock.Unlock()
+func (p Provisioner) mountDir(volID, targetPath string) error {
+	// Check if the target path is already mounted. Prevent remounting.
+	notMnt, err := mount.IsNotMountPoint(p.mounter, targetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(targetPath, 0750); err != nil {
+				return status.Error(codes.Internal, err.Error())
+			}
+			notMnt = true
+		} else {
+			return status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	if !notMnt {
+		return nil
+	}
+
+	if err := p.mounter.Mount(p.volPath(volID), targetPath, "", []string{"bind"}); err != nil {
+		var errList strings.Builder
+		errList.WriteString(err.Error())
+		if rmErr := os.RemoveAll(p.volPath(volID)); rmErr != nil && !os.IsNotExist(rmErr) {
+			errList.WriteString(fmt.Sprintf(" :%s", rmErr.Error()))
+		}
+		return status.Error(codes.Internal, fmt.Sprintf("failed to mount device: %s at %s: %s", p.volPath(volID), targetPath, errList.String()))
+	}
+
+	return nil
+}
 
-	if len(provisioner.Path) == 0 {
-		return "", fmt.Errorf("no base path provided")
+func (p Provisioner) writeFileToVolume(contents interface{}, volID, fileName string) error {
+	data, err := json.Marshal(contents)
+	if err != nil {
+		return status.Error(codes.Internal, fmt.Errorf("error marshalling file %s: %v", fileName, err).Error())
+	}
+	err = writeFile(data, filepath.Join(p.volPath(volID), fileName))
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
 	}
+	return nil
+}
 
-	if err := os.MkdirAll(filepath.Join(provisioner.Path, volumeID), 0755); err != nil {
-		return "", err
+func (p Provisioner) removeMount(path string) error {
+	err := mount.CleanupMountPoint(path, p.mounter, true)
+	if err != nil && !os.IsNotExist(err) {
+		klog.Error(err, "failed to clean and unmount target path", "targetPath", path)
+		return status.Error(codes.Internal, err.Error())
 	}
+	return nil
+}
 
-	return filepath.Join(provisioner.Path, volumeID), nil
+// fileExists returns whether the given file or directory exists
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
 }
 
-func Unprovision(vId string) error {
-	return os.RemoveAll(filepath.Join(provisioner.Path, vId))
+func writeFile(data []byte, filepath string) error {
+	klog.Infof("creating conn file: %s", filepath)
+
+	if exists, err := fileExists(filepath); exists || err != nil {
```

Comment:
why combine both checks here? Seems a bit more readable if separate, imo

Reply:
Separated the two checks out